zookeeper-metrics-providers/zookeeper-prometheus-metrics/src/main/java/org/apache/zookeeper/metrics/prometheus/PrometheusMetricsProvider.java

```diff
@@ -0,0 +1,295 @@
+ * limitations under the License.
+ */
+package org.apache.zookeeper.metrics.prometheus;
+
+import io.prometheus.client.Collector;
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.exporter.MetricsServlet;
+import io.prometheus.client.hotspot.DefaultExports;
+import java.util.Enumeration;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.BiConsumer;
+import org.apache.zookeeper.metrics.Counter;
+import org.apache.zookeeper.metrics.Gauge;
+import org.apache.zookeeper.metrics.MetricsContext;
+import org.apache.zookeeper.metrics.MetricsProvider;
+import org.apache.zookeeper.metrics.MetricsProviderLifeCycleException;
+import org.apache.zookeeper.metrics.Summary;
+import org.apache.zookeeper.metrics.SummarySet;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.servlet.ServletContextHandler;
+import org.eclipse.jetty.servlet.ServletHolder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A Metrics Provider implementation based on https://prometheus.io.
+ *
+ * @since 3.6.0
+ */
+public class PrometheusMetricsProvider implements MetricsProvider {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PrometheusMetricsProvider.class);
+    private final CollectorRegistry collectorRegistry = new CollectorRegistry(true);
+    private int port = 7000;
+    private Server server;
+    private final Context rootContext = new Context();
+
+    @Override
+    public void configure(Properties configuration) throws MetricsProviderLifeCycleException {
+        LOG.info("Initializing metrics, configuration: {}", configuration);
+        this.port = Integer.parseInt(configuration.getProperty("httpPort", "7000"));
+    }
+
+    @Override
+    public void start() throws MetricsProviderLifeCycleException {
+        try {
+            LOG.info("Starting /metrics HTTP endpoint at port " + port);
+            server = new Server(port);
+            ServletContextHandler context = new ServletContextHandler();
+            context.setContextPath("/");
+            server.setHandler(context);
+            context.addServlet(new ServletHolder(new MetricsServlet()), "/metrics");
+            server.start();
+        } catch (Exception err) {
+            LOG.error("Cannot start /metrics server", err);
+            server = null;
+            throw new MetricsProviderLifeCycleException(err);
+        }
+    }
+
+    @Override
+    public MetricsContext getRootContext() {
+        return rootContext;
+    }
+
+    @Override
+    public void stop() {
+        if (server != null) {
+            try {
+                server.stop();
+            } catch (Exception err) {
+                LOG.error("Cannot safely stop Jetty server", err);
+            } finally {
+                server = null;
+            }
+        }
+    }
+
+    @Override
+    public void dump(BiConsumer<String, Object> sink) {
```

Comment:
Why are you reimplementing the code in client_java? It correctly handles things like escaping.

Reply:
This is for the integration with "4 letter words commands" and the "Admin HTTP endpoint" (that is a JSON endpoint).
This is not consumed by Prometheus

Reply:
In that case, you're missing escaping.